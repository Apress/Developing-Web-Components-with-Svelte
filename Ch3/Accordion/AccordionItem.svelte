<svelte:options tag="cobalt-accordionitem" />

<script>
  import { slide } from "svelte/transition";
  import accordionData from "./accordiondata.json";
  export const data = accordionData;
  export let entry = "";
  export let title = "";
  let isOpen = false;
  const toggle = () => (isOpen = !isOpen);
</script>

<button on:click={toggle} aria-expanded={isOpen} class="accordionItem">
  <svg
    width="20"
    height="20"
    fill="none"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    stroke="currentColor"
    ><path d="M9 5l7 7-7 7" />
  </svg>
  {title}
</button>

{#if isOpen}
  <ul transition:slide={{ duration: 300 }}>
    <li>{entry}</li>
  </ul>
{/if}

<style>
  svg {
    transition: transform 0.2s ease-in;
  }

  [aria-expanded="true"] svg {
    transform: rotate(0.25turn);
  }

  button.accordionItem {
    display: flex;
    align-items: center;
    background-color: #6666ff;
    color: #ffffff;
    border: none;
  }

  button[aria-expanded="false"].accordionItem {
    margin-bottom: 2px;
  }

  button.accordionItem:hover {
    background-color: #19247c;
  }

  ul {
    border: 1px solid #6666ff;
    margin: 0;
    margin-bottom: 2px;
    padding: 20px 20px 20px 40px;
  }
</style>
