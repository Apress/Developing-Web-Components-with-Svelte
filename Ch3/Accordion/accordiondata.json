[
  { "title": "Heading 1", "text": "aaa" },
  { "title": "Heading 2", "text": "bbb" },
  { "title": "Heading 3", "text": "ccc" }
]