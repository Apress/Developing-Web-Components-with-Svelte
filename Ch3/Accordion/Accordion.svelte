<svelte:options tag="cobalt-accordion" />

<script>
	import AccordionItem from "./AccordionItem.svelte";
	//import accordionData from "./accordiondata.json";
	export let data = []; //= accordionData;
</script>

<div class="cobalt-accordion">
	{#each data as entry}
		<AccordionItem title={entry.title} entry={entry.text} />
	{/each}
</div>

<style>
	.cobalt-accordion {
		display: flex;
		flex-direction: column;
		width: 500px;
	}
</style>
