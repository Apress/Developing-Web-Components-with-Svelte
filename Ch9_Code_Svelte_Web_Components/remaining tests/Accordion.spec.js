import { render, fireEvent, screen } from '@testing-library/svelte';
import Accordion from "../src/lib/Accordion/Accordion.svelte";
import '@testing-library/jest-dom';

describe("Tests for Accordion", () => {
  const mockText = "This is a Accordion";

  const props = {
    data: [
      { "title": "Heading 1", "text": "aaa" },
      { "title": "Heading 2", "text": "bbb" },
      { "title": "Heading 3", "text": "ccc" }
    ]
  };

  it("should render properly", () => {
    const result = render(Accordion, { props });
    expect(() => result).not.toThrow();
  });

  it("should show the first heading with correct text", () => {
    const result = render(Accordion, props);
    expect(result).not.toBeNull();
  })

  // Should have text too, but throws a  null error
  it("should have the correct title", () => {
    render(Accordion, props); 
    
    expect(screen.queryByText(props.data[0].title)).toBeInTheDocument();
    expect(screen.queryByText(props.data[1].title)).toBeInTheDocument();
    expect(screen.queryByText(props.data[2].title)).toBeInTheDocument();
  })

  // https://testing-library.com/docs/svelte-testing-library/example
  it("should show aria-expanded as true when first item clicked", async() => {
    render(Accordion, props);  
    const AccordionObj = await(screen.getAllByRole("button")[0]);
    await fireEvent.click(AccordionObj);
    expect(screen.getAllByRole("button", {expanded: true})[0]).not.toBeNull();
  })

  it("get a snapshot of component", () => {
    const tree = render(Accordion, props);
    expect(tree).toMatchSnapshot();
  })
});