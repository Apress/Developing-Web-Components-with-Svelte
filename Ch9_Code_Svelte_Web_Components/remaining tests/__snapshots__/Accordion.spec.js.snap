// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Accordion get a snapshot of component 1`] = `
{
  "component": Accordion {
    "$$": {
      "after_update": [],
      "before_update": [],
      "bound": {},
      "callbacks": {},
      "context": Map {},
      "ctx": [
        [
          {
            "text": "aaa",
            "title": "Heading 1",
          },
          {
            "text": "bbb",
            "title": "Heading 2",
          },
          {
            "text": "ccc",
            "title": "Heading 3",
          },
        ],
      ],
      "dirty": [
        -1,
      ],
      "fragment": {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": [
        [Function],
      ],
      "on_disconnect": [],
      "on_mount": [],
      "props": {
        "data": 0,
      },
      "root": <div>
        <div
          class="cobalt-accordion svelte-1qnk47x"
        >
          <button
            aria-expanded="false"
            class="accordionItem svelte-yzpkvh"
          >
            <svg
              class="svelte-yzpkvh"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
            >
              <path
                d="M9 5l7 7-7 7"
              />
            </svg>
             
            Heading 1
          </button>
           
          
          <button
            aria-expanded="false"
            class="accordionItem svelte-yzpkvh"
          >
            <svg
              class="svelte-yzpkvh"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
            >
              <path
                d="M9 5l7 7-7 7"
              />
            </svg>
             
            Heading 2
          </button>
           
          
          <button
            aria-expanded="false"
            class="accordionItem svelte-yzpkvh"
          >
            <svg
              class="svelte-yzpkvh"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
            >
              <path
                d="M9 5l7 7-7 7"
              />
            </svg>
             
            Heading 3
          </button>
           
          
        </div>
      </div>,
      "skip_bound": false,
      "update": [Function],
    },
    "$$set": [Function],
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <div
        class="cobalt-accordion svelte-1qnk47x"
      >
        <button
          aria-expanded="false"
          class="accordionItem svelte-yzpkvh"
        >
          <svg
            class="svelte-yzpkvh"
            fill="none"
            height="20"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="20"
          >
            <path
              d="M9 5l7 7-7 7"
            />
          </svg>
           
          Heading 1
        </button>
         
        
        <button
          aria-expanded="false"
          class="accordionItem svelte-yzpkvh"
        >
          <svg
            class="svelte-yzpkvh"
            fill="none"
            height="20"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="20"
          >
            <path
              d="M9 5l7 7-7 7"
            />
          </svg>
           
          Heading 2
        </button>
         
        
        <button
          aria-expanded="false"
          class="accordionItem svelte-yzpkvh"
        >
          <svg
            class="svelte-yzpkvh"
            fill="none"
            height="20"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="20"
          >
            <path
              d="M9 5l7 7-7 7"
            />
          </svg>
           
          Heading 3
        </button>
         
        
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
